Infra: Docker images and status badges

HebertCL opened the pull request:
# TL;DR

Build workflow and README updates with badges for release and go mod.

## Summary

- Added `build-and-push.yml` workflow that builds and push Docker images to GitHub registry.
- Updated README mentioning the release section to reference the Docker images.
- Added README badges to reflect current Go mod version, build status (currently pointing to E2E for testing purposes) and release.

## Related issues

#12

commoddity commented on .github/workflows/build-and-push.yml:
```diff
@@ -0,0 +1,52 @@
+
+jobs:
+  publish-image:
+    runs-on: ubuntu-latest
```
Can we lock this to a version, eg. 22.04

HebertCL replied:
Great catch. I'm fixing this.

commoddity commented on README.md:
```diff
@@ -5,9 +5,16 @@
 </div>
 <br/>
 
+![GitHub Actions Workflow Status](https://img.shields.io/github/actions/workflow/status/buildwithgrove/path/run-e2e-tests.yml)
```
Love this.

commoddity commented on .github/workflows/build-and-push.yml:
```diff
@@ -0,0 +1,52 @@
+name: Publish image to GitHub Container Registry
+
+on:
```
I think this workflow should either run the tests itself or be triggered a successful run of the test workflow. WDYT?

HebertCL replied:
I agree with this. No sense in building an image from a commit where tests are failing. Question would be: Do we want to chain this to run right after e2e, or do we want it to build it only by running unit testing?

commoddity replied:
My preference would be have a CI job that does all our checks: lint, unit test, E2E test, test Docker build, etc. then triggers this workflow if it all succeeds.

HebertCL replied:
I've taken the workflows from `portal-middleware` to enhance this.

commoddity commented on .github/workflows/build-and-push.yml:
```diff
@@ -0,0 +1,52 @@
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Docker Metadata action
```
Am I understanding this right that this tags the image based on the Github version tags we create manually on release? 

If so that seems sufficient for now rather than implementing automated versioning, although automating the semver tag creation is also an option.

HebertCL replied:
I merely ripped off this metadata config from poktroll repo. I haven't been able to see that your statement is what it does (like you, I assume that's how it works). I also think automating semver is an good option, so I would say whether we go for one or the other, we should be just fine.

commoddity reviewed (commented):
Great PR, just a few questions.

Olshansk commented on .github/workflows/build-and-push.yml:
```diff
@@ -0,0 +1,55 @@
+          username: ${{ github.actor }}
+          password: ${{ github.token }}
+
+      ## TODO: Do we want/need to make PATH multi arch?
```
Yes please, can you do it in this PR?

PATH is something I'd like to be able to run on my M1 processor.

Hit a lot of issues with in pocketdex so let's just get ahead of it: https://github.com/pokt-network/pocketdex/blob/1178ca47f68f6596bed7658dc1ba2106cdda1362/.github/workflows/ci-docker.yml#L33

cc @bryanchriswhite @okdas for visibility

okdas replied:
Yup. Also, more providers (including the lean services such as hetzner) add arm64 to their offerings, so I say we need it.

HebertCL replied:
Added ARM arch for docker build.

Olshansk commented on .github/workflows/build-and-push.yml:
```diff
@@ -0,0 +1,55 @@
+            type=sha,suffix=-rc
+            type=sha,format=long,suffix=-rc
+
+      ## TODO: Is the GH token in place in the repo?
```
This seems like a `TODO_IN_THIS_PR`.

See https://github.com/pokt-network/poktroll/blob/ecf74ced63cc920181827ebfb04bf03853fff210/.github/workflows/reviewdog.yml#L12

HebertCL replied:
I've solved this doubt thanks to @okdas

Olshansk commented on .github/workflows/run-e2e-tests.yml:
```diff
@@ -2,12 +2,16 @@ name: Run E2E Tests
 
 on:
   push:
+    # paths-ignore:
```
remove?

HebertCL replied:
Removed.

Olshansk reviewed (commented):
Let's go with multi-arch and make sure we resolve the token issue in this PR.

Also, I usually use this to test github workflows locally: https://github.com/nektos/act

(Just in case you didn't know about it)

okdas reviewed (changes requested):
Looks great! I'm actually already using an image built by this PR in some load-testing work. ❤️ There are a couple of things I think we should do before we merge this.

1. `gocritic` is part of `golangci-lint`. You can enable it in `.golangci.yml` and remove the separate job (saves CI time):

```yaml
linters:
  disable-all: true
  enable:
    - gocritic
```
You can also change gocritic params that way: https://golangci-lint.run/usage/linters/#gocritic

2. It appears `golangci-lint` produces some errors in the CI. It would be great if they are addressed in the PR as well so they don't make it into other people's branches. Please let me know if you need assistance there.

commoddity commented on .github/workflows/test-build-push.yml:
```diff
@@ -0,0 +1,106 @@
+
+      - name: Run unit tests
+        run: CGO_ENABLED=0 go test ./... -short
+
```
Any reason we can't just run the E2E test job in this workflow and remove the separate file?

commoddity replied:
OK fine with me. I generally prefer waiting a bit longer from more confidence from more comprehensive tests but I'm fine with keeping them separate. I can always just run the E2E locally.

HebertCL replied:
I've arranged the jobs in a way where unit testing and e2e run in parallel, but the build job requires both of them to be successful before running the build. That would give us confidence that both types of tests get executed, taking less CI time and ensuring that any of them failing will not produce a non-functional image.

commoddity commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,17 @@
+
+on:
+  push:
+    branches: ["main"]
```
I think this should be triggered by a successful run of `Run tests, build image and push to GitHub Container Registry`.

Olshansk commented on .github/workflows/release.yml:
```diff
@@ -0,0 +1,17 @@
+
+jobs:
+  release-on-push:
+    runs-on: ubuntu-22.04
```
@okdas For consistency, is this what you're using now too?

HebertCL replied:
AFAICT poktroll is using `ubuntu-latest`. I agree with @commoddity tho that it is a better approach to fix the runner version. In this case, I'm using the same fixed version used in middleware.

See for reference:
https://github.com/pokt-network/poktroll/blob/3384138248474a2e724445384fbaff07ab0c41b2/.github/workflows/main-build.yml#L14
https://github.com/pokt-network/poktroll/blob/3384138248474a2e724445384fbaff07ab0c41b2/.github/workflows/release-artifacts.yml#L14

Olshansk commented on .github/workflows/test-build-push.yml:
```diff
@@ -0,0 +1,125 @@
+name: Run tests, build image and push to GitHub Container Registry
```
Any chance we can follow the same pattern / split as in `poktroll`?

Specifically, we have a `main-build.yml`, a `release-artifacts.yml` a `run-tests.yml`.

I realize it's a little overkill, but don't think it should be too much overhead here.

https://github.com/pokt-network/poktroll/tree/main/.github/workflows

HebertCL replied:
We can follow this pattern for the sake of consistency when it comes to files and naming convention. I'll take a look at the files and see what we can adapt (I've already taken parts of them).

HebertCL replied:
Updated to reflect the suggested pattern

Olshansk commented on README.md:
```diff
@@ -24,6 +35,10 @@
 
+## Path Releases
+
+Path releases provide a Docker image you can start using right away to bootstrap your Path gateway without the need of building your own image. Head over to [releases](https://github.com/buildwithgrove/path/releases) to download it.
```
Can you provide an example command on how to download the docker image?

#PUC (Please Update Comment)

HebertCL replied:
This section has been updated.

Olshansk reviewed (changes requested):
Please re-reqest a review once it's ready.

HebertCL commented:
> Looks great! I'm actually already using an image built by this PR in some load-testing work. ❤️ There are a couple of things I think we should do before we merge this.
> 
> 1. `gocritic` is part of `golangci-lint`. You can enable it in `.golangci.yml` and remove the separate job (saves CI time):
> 
> ```yaml
> linters:
>   disable-all: true
>   enable:
>     - gocritic
> ```
> 
> You can also change gocritic params that way: https://golangci-lint.run/usage/linters/#gocritic
> 
> 2. It appears `golangci-lint` produces some errors in the CI. It would be great if they are addressed in the PR as well so they don't make it into other people's branches. Please let me know if you need assistance there.

I've addressed this using a `golangci.yml` file that now the job is referring to for execution.

commoddity commented on .github/workflows/release-artifacts.yml:
```diff
@@ -0,0 +1,17 @@
+name: Tag and Release
+
+on:
```
Shouldn't this one also only run when the tests pass?

HebertCL replied:
This is true. We're pursuing to automate this release part and not tie it to a release we manually create. I'll update this workflow to reflect it.

HebertCL replied:
This has been addressed.

commoddity commented on .github/workflows/main-build.yml:
```diff
@@ -0,0 +1,50 @@
+  workflow_run:
+    workflows: ["Path tests"]
+    types:
+      - completed
```
I think this needs to specify a successful run, not just completed.

eg:
```yaml
on:
  workflow_run:
    workflows: ["Path tests"]
    types:
      - completed
    if: ${{ github.event.workflow_run.conclusion == 'success' }}
    ```

HebertCL replied:
This has been addressed.

commoddity commented on .github/workflows/run-tests.yml:
```diff
@@ -0,0 +1,81 @@
+      - name: Run golangci-lint
+        uses: golangci/golangci-lint-action@v6
+        with:
+          version: latest
```
I wonder if we should peg this to a set version as well?

HebertCL replied:
I think it's ok if we keep using `latest` since it's the golangci version and we're not pairing this with specific golang installation.

commoddity commented on .github/workflows/main-build.yml:
```diff
@@ -0,0 +1,50 @@
+      - name: Build and push Docker image
+        uses: docker/build-push-action@v5
+        with:
+          push: true
```
Can we add the tag as a build arg?

eg.
```yaml
- name: Build and push Docker image
  uses: docker/build-push-action@v5
  with:
    push: true
    tags: ${{ steps.meta.outputs.tags }}
    build-args: IMAGE_TAG=${{ steps.meta.outputs.tags }}
....
```

HebertCL replied:
This has been addressed.

commoddity commented:
> > Looks great! I'm actually already using an image built by this PR in some load-testing work. ❤️ There are a couple of things I think we should do before we merge this.
> > 
> > 1. `gocritic` is part of `golangci-lint`. You can enable it in `.golangci.yml` and remove the separate job (saves CI time):
> > 
> > ```yaml
> > linters:
> >   disable-all: true
> >   enable:
> >     - gocritic
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can also change gocritic params that way: https://golangci-lint.run/usage/linters/#gocritic
> > 
> > 2. It appears `golangci-lint` produces some errors in the CI. It would be great if they are addressed in the PR as well so they don't make it into other people's branches. Please let me know if you need assistance there.
> 
> I've addressed this using a `golangci.yml` file that now the job is referring to for execution.

Super minor NIT but there's a typo in the filename: it's `.golanci.yml` instead of `.golangci.yml`.

commoddity reviewed (changes requested):
Just a reminder on the comments I added; forgot to hit `Request Changes`.

HebertCL commented:
> > > Looks great! I'm actually already using an image built by this PR in some load-testing work. ❤️ There are a couple of things I think we should do before we merge this.
> > > 
> > > 1. `gocritic` is part of `golangci-lint`. You can enable it in `.golangci.yml` and remove the separate job (saves CI time):
> > > 
> > > ```yaml
> > > linters:
> > >   disable-all: true
> > >   enable:
> > >     - gocritic
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > You can also change gocritic params that way: https://golangci-lint.run/usage/linters/#gocritic
> > > 
> > > 2. It appears `golangci-lint` produces some errors in the CI. It would be great if they are addressed in the PR as well so they don't make it into other people's branches. Please let me know if you need assistance there.
> > 
> > 
> > I've addressed this using a `golangci.yml` file that now the job is referring to for execution.
> 
> Super minor NIT but there's a typo in the filename: it's `.golanci.yml` instead of `.golangci.yml`.

🤦🏻 no wonder my linter was failing. I've fixed this.

commoddity reviewed (approved):
LGTM

okdas reviewed (approved):
Looks great, thank you!

I've had issues with `on.workflow_run` in the past, and curious how that'll work with releases!
